zone/raids.cpp

```diff
@@ -1884,3 +2025,84 @@ bool Raid::DoesAnyMemberHaveExpeditionLockout(
 		return Expedition::HasLockoutByCharacterName(raid_member.membername, expedition_name, event_name);
 	});
 }
+
+Mob* Raid::GetRaidMainAssistOneByName(const char* name)
+{
+	Raid* raid = entity_list.GetRaidByBotName(name);
+	std::vector<RaidMember> raid_members = raid->GetMembers();
+
+	for (RaidMember iter : raid_members)
+	{
+		if (iter.IsRaidMainAssistOne) {
+			return iter.member->CastToMob();
+		}
+	}
+	return nullptr;
+}
+bool Raid::IsEngaged() {
+
+	std::vector<RaidMember> rm = GetMembers();
+	for (const auto& m : rm) {
+		if (m.member && !m.IsBot && (m.member->IsEngaged() || m.member->GetAggroCount() > 0)) {
+			return 1;
+		}
+	}
+	return 0;
+}
+void Raid::RaidGroupSay(const char* msg, const char* from, uint8 language, uint8 lang_skill)
+{
+	if (!from)
+		return;
+
+	uint32 groupToUse = GetGroup(from);
+
+	if (groupToUse > 11)
+		return;
+
+	auto pack = new ServerPacket(ServerOP_RaidGroupSay, sizeof(ServerRaidMessage_Struct) + strlen(msg) + 1);
+	ServerRaidMessage_Struct* rga = (ServerRaidMessage_Struct*)pack->pBuffer;
+	rga->rid = GetID();
+	rga->gid = groupToUse;
+	rga->language = language;
+	rga->lang_skill = lang_skill;
+	strn0cpy(rga->from, from, 64);
+
+	strcpy(rga->message, msg); // this is safe because we are allocating enough space for the entire msg above
+
+	worldserver.SendPacket(pack);
+	safe_delete(pack);
+}
+void Raid::RaidSay(const char* msg, const char* from, uint8 language, uint8 lang_skill)
+{
+	if (!from)
+		return;
+
+	auto pack = new ServerPacket(ServerOP_RaidSay, sizeof(ServerRaidMessage_Struct) + strlen(msg) + 1);
+	ServerRaidMessage_Struct* rga = (ServerRaidMessage_Struct*)pack->pBuffer;
+	rga->rid = GetID();
+	rga->gid = RAID_GROUPLESS;
+	rga->language = language;
+	rga->lang_skill = lang_skill;
+	strn0cpy(rga->from, from, 64);
+
+	strcpy(rga->message, msg);
+
+	worldserver.SendPacket(pack);
+	safe_delete(pack);
+}
+
+void Raid::SetNewRaidLeader(uint32 i)
+{
+	if (members[i].IsRaidLeader) {
+		for (int x = 0; x < MAX_RAID_MEMBERS; x++) {
```

Comment:
You can range through the array versus doing the integer iteration

eg

```
for (auto & m : members) {
```

Accessing

```cpp
m.member->DoThing()
```